ring/ownership_priority_queue.go

```diff
@@ -0,0 +1,200 @@
+// order of item ids.
+type ownershipPriorityQueue struct {
+	items []*ownershipInfo
+}
+
+func newPriorityQueue(len int) *ownershipPriorityQueue {
+	return &ownershipPriorityQueue{
+		items: make([]*ownershipInfo, 0, len),
+	}
+}
+
+func (pq *ownershipPriorityQueue) Len() int {
+	return len(pq.items)
+}
+
+func (pq *ownershipPriorityQueue) Swap(i, j int) {
+	pq.items[i], pq.items[j] = pq.items[j], pq.items[i]
+	pq.items[i].index = i // pq[i] is actually the old pq[j], so pq[i].index should be updated to i
+	pq.items[j].index = j // pq[j] is actually the old pq[i], so pq[j].index should be updated to j
+}
+
+func (pq *ownershipPriorityQueue) Less(i, j int) bool {
+	if pq.items[i].ownership == pq.items[j].ownership {
+		// In order to guarantee the stability, i.e., that the same instanceID and zone as input
+		// always generate the same slice of tokens as output, we enforce that by equal ownership
+		// higher priority is determined by the order of ids.
+		return pq.compareRingItems(pq.items[i].ringItem, pq.items[j].ringItem)
+	}
+	// We are implementing a max-heap, so we are using > here.
+	// Since we compare float64, NaN values must be placed at the end.
+	return pq.items[i].ownership > pq.items[j].ownership || (math.IsNaN(pq.items[j].ownership) && !math.IsNaN(pq.items[i].ownership))
+}
+
+func (pq *ownershipPriorityQueue) compareRingItems(ri1, ri2 ringItem) bool {
+	if ri1 == nil || ri2 == nil {
+		return true
+	}
+	// we invert the order because we want that instances with higher ids have a higher priority
+	return ri2.Less(ri1)
+}
+
+// Push implements heap.Push(any). It pushes the element item onto ownershipPriorityQueue.
+// Time complexity is O(log n), where n = Len().
+func (pq *ownershipPriorityQueue) Push(item any) {
+	n := len(pq.items)
+	ownershipInfo := item.(*ownershipInfo)
+	ownershipInfo.index = n
+	pq.items = append(pq.items, ownershipInfo)
+}
+
+// Pop implements heap.Pop(). It removes and returns the element with the highest priority from ownershipPriorityQueue.
+// Time complexity is O(log n), where n = Len().
+func (pq *ownershipPriorityQueue) Pop() any {
+	old := pq.items
+	n := len(old)
+	item := old[n-1]
+	old[n-1] = nil
+	item.index = -1
+	pq.items = old[0 : n-1]
+	return item
+}
+
+// Peek the returns the element with the highest priority from ownershipPriorityQueue,
+// but it does not remove it from the latter. Time complexity is O(1).
+func (pq *ownershipPriorityQueue) Peek() *ownershipInfo {
+	return (pq.items)[0]
+}
+
+// Update updates the element ownershipInfo passed as parameter by applying to it the updating function update
+// passed as parameter, and propagates this modification to ownershipPriorityQueue. Element ownershipInfo must
+// be already present on ownershipPriorityQueue. Time complexity is O(log n), where n = Len().
+func (pq *ownershipPriorityQueue) Update(ownershipInfo *ownershipInfo, update func(*ownershipInfo)) {
+	update(ownershipInfo)
+	heap.Fix(pq, ownershipInfo.index)
```

Comment:
The fact that `ownershipInfo` had an `index` was bothering me, so I followed the code and I don't think that it's needed, neither this `Update` method.
The only two usages of this method use an info that came from peek, so it's index is always 0. I tried adding this here:

```
	if ownershipInfo.index != 0 {
		panic("HERE IT IS NOT 0" + strconv.Itoa(ownershipInfo.index))
	}
```

And the only test that failed was the `Update` one (because it violates that assumption). 

I would remove `index` from `ownershipPriorityQueue` and all the logic related to keeping track of it. Then instead of calling `Update` with a lambda, I'd just update the element and call `heap.Fix(pq, 0)`.